File: utils/analytics-collector/python/aggregate.py
"""Aggregation functions for player analytics (stdlib only)."""

from typing import Any


def compute_path_stats(entries: list[dict[str, Any]]) -> dict[str, Any]:
    """
    Compute path statistics from logged events.

    Args:
        entries: Array of log entries

    Returns:
        Dict with totalChoices, uniqueSessions, pathDistribution
    """
    choices = [e for e in entries if e.get("type") == "choice"]

    # Group by knotPath -> choiceIndex -> { count, text }
    by_path: dict[str, dict[int, dict[str, Any]]] = {}
    sessions: set[str] = set()

    for c in choices:
        sessions.add(c.get("sessionId", ""))

        payload = c.get("payload") or {}
        path = payload.get("knotPath", "unknown")
        choice_index = payload.get("choiceIndex", -1)

        if path not in by_path:
            by_path[path] = {}
        if choice_index not in by_path[path]:
            by_path[path][choice_index] = {
                "count": 0,
                "text": payload.get("choiceText", ""),
            }
        by_path[path][choice_index]["count"] += 1

    # Convert int keys to strings for JSON compatibility
    path_distribution = {
        path: {str(idx): data for idx, data in choices_map.items()}
        for path, choices_map in by_path.items()
    }

    return {
        "totalChoices": len(choices),
        "uniqueSessions": len(sessions),
        "pathDistribution": path_distribution,
    }


def compute_session_summaries(entries: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """
    Compute session summaries.

    Args:
        entries: Array of log entries

    Returns:
        List of session summaries
    """
    sessions: dict[str, dict[str, Any]] = {}

    for entry in entries:
        sid = entry.get("sessionId", "")
        timestamp = entry.get("timestamp", 0)

        if sid not in sessions:
            sessions[sid] = {
                "sessionId": sid,
                "startTime": timestamp,
                "endTime": timestamp,
                "choiceCount": 0,
                "paths": set(),
            }

        s = sessions[sid]
        s["startTime"] = min(s["startTime"], timestamp)
        s["endTime"] = max(s["endTime"], timestamp)

        if entry.get("type") == "choice":
            s["choiceCount"] += 1
            payload = entry.get("payload") or {}
            knot_path = payload.get("knotPath")
            if knot_path:
                s["paths"].add(knot_path)

    return [
        {
            "sessionId": s["sessionId"],
            "startTime": s["startTime"],
            "endTime": s["endTime"],
            "choiceCount": s["choiceCount"],
            "paths": list(s["paths"]),
            "durationMs": s["endTime"] - s["startTime"],
        }
        for s in sessions.values()
    ]


def top_paths(entries: list[dict[str, Any]], limit: int = 10) -> list[dict[str, Any]]:
    """
    Find most common choice paths.

    Args:
        entries: Array of log entries
        limit: Number of top paths to return

    Returns:
        Top paths by frequency
    """
    choices = [e for e in entries if e.get("type") == "choice"]
    path_counts: dict[str, int] = {}

    for c in choices:
        payload = c.get("payload") or {}
        path = payload.get("knotPath", "unknown")
        path_counts[path] = path_counts.get(path, 0) + 1

    sorted_paths = sorted(path_counts.items(), key=lambda x: x[1], reverse=True)

    return [{"path": path, "count": count} for path, count in sorted_paths[:limit]]


def compute_choice_stats(
    entries: list[dict[str, Any]], choice_ids: list[str]
) -> dict[str, Any]:
    """
    Compute statistics for specific choice IDs (Telltale-style).

    Choice ID format: "knotName:choiceIndex" (e.g., "interrogation:0")

    Args:
        entries: Array of log entries
        choice_ids: List of choice IDs to query

    Returns:
        Dict with choices mapping to total and percentage
    """
    choices = [e for e in entries if e.get("type") == "choice"]

    # Count totals per knot for percentage calculation
    knot_totals: dict[str, int] = {}
    choice_counts: dict[str, dict[str, Any]] = {}

    for c in choices:
        payload = c.get("payload") or {}
        context = c.get("context") or {}
        knot = payload.get("knotPath") or context.get("knot") or "unknown"
        idx = payload.get("choiceIndex", -1)
        key = f"{knot}:{idx}"

        # Count per knot
        knot_totals[knot] = knot_totals.get(knot, 0) + 1

        # Count per choice
        if key not in choice_counts:
            choice_counts[key] = {"count": 0, "knot": knot}
        choice_counts[key]["count"] += 1

    # Build response for requested choice IDs
    result: dict[str, dict[str, int]] = {}
    for choice_id in choice_ids:
        knot = choice_id.split(":")[0] if ":" in choice_id else choice_id
        data = choice_counts.get(choice_id)

        if data:
            total = knot_totals.get(knot, 1)
            result[choice_id] = {
                "total": data["count"],
                "percentage": round((data["count"] / total) * 100),
            }
        else:
            result[choice_id] = {"total": 0, "percentage": 0}

    return {"choices": result}

File: utils/analytics-collector/node/aggregate.js
// Aggregation functions for player analytics

/**
 * Compute path statistics from logged events
 * @param {Array} entries - Array of log entries
 * @returns {Object} - { totalChoices, uniqueSessions, pathDistribution }
 */
export function computePathStats(entries) {
  const choices = entries.filter((e) => e.type === 'choice');

  // Group by knotPath -> choiceIndex -> { count, text }
  const byPath = {};
  const sessions = new Set();

  for (const c of choices) {
    sessions.add(c.sessionId);

    const path = c.payload?.knotPath || 'unknown';
    const choiceIndex = c.payload?.choiceIndex ?? -1;

    byPath[path] ??= {};
    byPath[path][choiceIndex] ??= {
      count: 0,
      text: c.payload?.choiceText || '',
    };
    byPath[path][choiceIndex].count++;
  }

  return {
    totalChoices: choices.length,
    uniqueSessions: sessions.size,
    pathDistribution: byPath,
  };
}

/**
 * Compute session summaries
 * @param {Array} entries - Array of log entries
 * @returns {Array} - Array of session summaries
 */
export function computeSessionSummaries(entries) {
  const sessions = {};

  for (const entry of entries) {
    const sid = entry.sessionId;
    sessions[sid] ??= {
      sessionId: sid,
      startTime: entry.timestamp,
      endTime: entry.timestamp,
      choiceCount: 0,
      paths: new Set(),
    };

    const s = sessions[sid];
    s.startTime = Math.min(s.startTime, entry.timestamp);
    s.endTime = Math.max(s.endTime, entry.timestamp);

    if (entry.type === 'choice') {
      s.choiceCount++;
      if (entry.payload?.knotPath) {
        s.paths.add(entry.payload.knotPath);
      }
    }
  }

  return Object.values(sessions).map((s) => ({
    ...s,
    paths: [...s.paths],
    durationMs: s.endTime - s.startTime,
  }));
}

/**
 * Find most common choice paths
 * @param {Array} entries - Array of log entries
 * @param {number} limit - Number of top paths to return
 * @returns {Array} - Top paths by frequency
 */
export function topPaths(entries, limit = 10) {
  const choices = entries.filter((e) => e.type === 'choice');
  const pathCounts = {};

  for (const c of choices) {
    const path = c.payload?.knotPath || 'unknown';
    pathCounts[path] ??= 0;
    pathCounts[path]++;
  }

  return Object.entries(pathCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, limit)
    .map(([path, count]) => ({ path, count }));
}

/**
 * Compute statistics for specific choice IDs (Telltale-style)
 * Choice ID format: "knotName:choiceIndex" (e.g., "interrogation:0")
 *
 * @param {Array} entries - Array of log entries
 * @param {string[]} choiceIds - Array of choice IDs to query
 * @returns {Object} - { choices: { "knot:0": { total, percentage } } }
 */
export function computeChoiceStats(entries, choiceIds) {
  const choices = entries.filter((e) => e.type === 'choice');

  // Count totals per knot for percentage calculation
  const knotTotals = {};
  const choiceCounts = {};

  for (const c of choices) {
    const knot = c.payload?.knotPath || c.context?.knot || 'unknown';
    const idx = c.payload?.choiceIndex ?? -1;
    const key = `${knot}:${idx}`;

    // Count per knot
    knotTotals[knot] ??= 0;
    knotTotals[knot]++;

    // Count per choice
    choiceCounts[key] ??= { count: 0, knot };
    choiceCounts[key].count++;
  }

  // Build response for requested choice IDs
  const result = {};
  for (const id of choiceIds) {
    const [knot] = id.split(':');
    const data = choiceCounts[id];

    if (data) {
      const total = knotTotals[knot] || 1;
      result[id] = {
        total: data.count,
        percentage: Math.round((data.count / total) * 100),
      };
    } else {
      result[id] = { total: 0, percentage: 0 };
    }
  }

  return { choices: result };
}
